array.c

```diff
@@ -4373,65 +4373,99 @@ take_items(VALUE obj, long n)
- *  whose elements are Arrays.
+ *  With no block given, combines +self+ with a collection of other arrays;
+ *  returns a new array of sub-arrays.
  *
- *  Each nested array <tt>new_array[n]</tt> is of size <tt>other_arrays.size+1</tt>,
- *  and contains:
+ *  In a common case, all the +objects+ are arrays,
+ *  and so the "other arrays" are just the +objects+ themselves;
+ *  for other cases (see below), each non-array object is converted to an array
+ *  that becomes one of the "other arrays."
+ *
+ *  The returned object is an array of size <tt>objects.size + 1</tt>
+ *  (that is, the count of other arrays plus one), and contains:
  *
  *  - The _nth_ element of +self+.
- *  - The _nth_ element of each of the +other_arrays+.
+ *  - The _nth_ element of each of the other arrays, as available.
  *
- *  If all +other_arrays+ and +self+ are the same size:
+ *  When the other arrays are all the same size as +self+,
+ *  the returned sub-arrays are a rearrangement containing exactly elements of all the arrays
+ *  (including +self+), with no omissions or additions:
  *
  *    a = [:a0, :a1, :a2, :a3]
  *    b = [:b0, :b1, :b2, :b3]
  *    c = [:c0, :c1, :c2, :c3]
  *    d = a.zip(b, c)
- *    d # => [[:a0, :b0, :c0], [:a1, :b1, :c1], [:a2, :b2, :c2], [:a3, :b3, :c3]]
+ *    pp d
+ *    # =>
+ *    [[:a0, :b0, :c0],
+ *     [:a1, :b1, :c1],
+ *     [:a2, :b2, :c2],
+ *     [:a3, :b3, :c3]]
  *
- *  If any array in +other_arrays+ is smaller than +self+,
- *  fills to <tt>self.size</tt> with +nil+:
+ *  When one of the other arrays is smaller than +self+,
+ *  pads the corresponding sub-array with +nil+ elements:
  *
  *    a = [:a0, :a1, :a2, :a3]
  *    b = [:b0, :b1, :b2]
  *    c = [:c0, :c1]
  *    d = a.zip(b, c)
- *    d # => [[:a0, :b0, :c0], [:a1, :b1, :c1], [:a2, :b2, nil], [:a3, nil, nil]]
+ *    pp d
+ *    # =>
+ *    [[:a0, :b0, :c0],
+ *     [:a1, :b1, :c1],
+ *     [:a2, :b2, nil],
+ *     [:a3, nil, nil]]
  *
- *  If any array in +other_arrays+ is larger than +self+,
- *  its trailing elements are ignored:
+ *  When one of the other arrays is larger than +self+,
+ *  ignores(!) its trailing elements:
  *
  *    a = [:a0, :a1, :a2, :a3]
  *    b = [:b0, :b1, :b2, :b3, :b4]
  *    c = [:c0, :c1, :c2, :c3, :c4, :c5]
  *    d = a.zip(b, c)
- *    d # => [[:a0, :b0, :c0], [:a1, :b1, :c1], [:a2, :b2, :c2], [:a3, :b3, :c3]]
- *
- *  If an argument is not an array, it extracts the values by calling #each:
- *
- *    a = [:a0, :a1, :a2, :a2]
- *    b = 1..4
- *    c = a.zip(b)
- *    c # => [[:a0, 1], [:a1, 2], [:a2, 3], [:a2, 4]]
- *
- *  When a block is given, calls the block with each of the sub-arrays (formed as above); returns +nil+:
- *
+ *    pp d
+ *    # =>
+ *    [[:a0, :b0, :c0],
+ *     [:a1, :b1, :c1],
+ *     [:a2, :b2, :c2],
+ *     [:a3, :b3, :c3]]
+ *
+ *  For an *object* that is _not_ an array,
+ *  forms the the "other array" as <tt>object.each.to_a</tt>:
```

Comment:
This is not true. It will first attempt to convert using `to_ary`, and then fall back to `each.to_a`. For example:

```ruby
obj_with_to_a = Object.new
def obj_with_to_a.to_ary
  puts "obj_with_to_a#to_a called!"
  [1, 2, 3]
end

obj_with_each = Object.new
def obj_with_each.each
  puts "obj_with_each#each called!"
  yield "one"
  yield "two"
  yield "three"
end

p [0, 0, 0].zip(obj_with_to_a, obj_with_each)
# => obj_with_to_a#to_a called!
# => obj_with_each#each called!
# => [[0, 1, "one"], [0, 2, "two"], [0, 3, "three"]]

Reply:
Recast.